packages/venia-ui/lib/components/CheckoutPage/PaymentInformation/brainTreeDropIn.js

```diff
@@ -0,0 +1,173 @@
+                            required: true
+                        },
+                        overrides: {
+                            fields: {
+                                number: {
+                                    maskInput: {
+                                        // Only show last four digits on blur.
+                                        showLastFour: true
+                                    }
+                                }
+                            }
+                        }
+                    }
+                });
+
+                if (didClose) {
+                    // If we get here but the form is closed we should teardown
+                    // instead of trying to use the instance.
+                    dropinInstance.teardown();
+                } else {
+                    setDropinInstance(dropinInstance);
+                    onReady(true);
+                }
+            } catch (err) {
+                console.error(
+                    `Unable to initialize Credit Card form (Braintree). \n${err}`
+                );
+
+                if (!didClose) {
+                    // Error state used to render error view. If we got an error
+                    // but we are closing the form we shouldn't try to update
+                    // state.
+                    setIsError(true);
+                }
+            }
+        },
+        [onReady, containerId]
+    );
+
+    useEffect(() => {
+        let didClose = false;
+
+        // Initialize the dropin with a reference to the container mounted in
+        // this component. We do this via onmount because we have to be sure
+        // the container id has mounted per braintree's API.
+        createDropinInstance(didClose);
+
+        // If we close the form quickly after opening we may end up in a
+        // semi-mounted state so we need a local variable to make sure we
+        // behave correctly.
+        return () => {
+            didClose = true;
+        };
+    }, [createDropinInstance]);
+
+    useEffect(() => {
+        async function requestPaymentNonce() {
+            try {
+                const paymentNonce = await dropinInstance.requestPaymentMethod();
+                onSuccess(paymentNonce);
+            } catch (e) {
+                // An error occurred. BrainTree will update the UI with error
+                // messaging, but we should signal that there was an error.
+                console.error(`Invalid Payment Details. \n${e}`);
+                onError();
+            }
+        }
+
+        if (shouldRequestPaymentNonce) {
+            requestPaymentNonce();
+        }
+    }, [dropinInstance, onError, onSuccess, shouldRequestPaymentNonce]);
+
+    useEffect(() => {
+        if (shouldTeardownDropin) {
+            dropinInstance
+                .teardown()
+                .then(() => {
+                    resetShouldTeardownDropin();
+                    createDropinInstance();
```

Comment:
What is the difference between this effect and [this teardown on line 71](https://github.com/magento/pwa-studio/pull/2320/files#diff-85ae277ce64dc8b57308b14d64ea72f1R71)?

Reply:
I am not sure of the internals of the dropping, but Andy added it as part of the initial drop in work that he did for the mini cart. Here I had to move it into a diff function because of a case that can arise due to payment service failing.

Here is some background information.

The credit card checkout process is a 3 step process:
1. Save billing address information on the cart
2. Get payment nonce from braintree by submitting credit card information
3. Save the payment nonce that we got from braintree in step 2.

After step 2, the braintree dropin changes it UI from a form to a summary kind of component which does not confine with out UX.

If the save service in step 3 fails, the dropin needs to be reset to the form. This is where tear down comes into picture. When that happens, we send a signal to the braintree dropin to teardown, which in braintree terms is reset so the user can enter the new payment information and submit again.

Reply:
The reason I asked was to try to get an understanding of both cases and see if there was some simplification we could make. In this instance it looks like the original `.teardown()` call was used as a sort of cleanup where the modal/dropin was unmounted. That doesn't seem to need the same logic as you've done in your async calls so there's nothing to do here.

There is an edge case I can forsee causing some errors be thrown. If `shouldTeardownDropin` is `true` the effect will run triggering teardown/recreation of the dropin instance. If a user closes the form during this effect's execution an error will throw because the dropin will mount without a target. That's why we added the `didClose` logic in the past.